server/s3_copy_object_action.cc

```diff
@@ -282,26 +274,294 @@ void S3CopyObjectAction::fetch_object_info_success() {
+
+  content_length = source_object_metadata->get_content_length();
+  _set_layout_id(S3MotrLayoutMap::get_instance()->get_layout_for_object_size(
+      content_length));
+  motr_unit_size =
+      S3MotrLayoutMap::get_instance()->get_unit_size_for_layout(layout_id);
+  assert(motr_unit_size > 0);
+  // TODO: handle tried_count
+
+  if (!motr_writer) {
+    motr_writer =
+        motr_writer_factory->create_motr_writer(request, new_object_oid);
+  } else {
+    motr_writer->set_oid(new_object_oid);
+  }
+  motr_writer->create_object(
+      std::bind(&S3CopyObjectAction::create_object_successful, this),
+      std::bind(&S3CopyObjectAction::create_object_failed, this), layout_id);
+
+  // for shutdown testcases, check FI and set shutdown signal
+  S3_CHECK_FI_AND_SET_SHUTDOWN_SIGNAL(
+      "put_object_action_create_object_shutdown_fail");
+  s3_log(S3_LOG_DEBUG, "", "Exiting\n");
+}
+
+void S3CopyObjectAction::create_object_successful() {
+  s3_log(S3_LOG_INFO, request_id, "Entering");
+  s3_copy_action_state = S3CopyObjectActionState::newObjOidCreated;
+
+  // New Object or overwrite, create new metadata and release old.
+  new_object_metadata.reset(new S3ObjectMetadata(*source_object_metadata));
+
+  new_object_metadata->set_object_list_index_oid(
+      bucket_metadata->get_object_list_index_oid());
+
+  new_object_metadata->set_objects_version_list_index_oid(
+      bucket_metadata->get_objects_version_list_index_oid());
+
+  new_oid_str = S3M0Uint128Helper::to_string(new_object_oid);
+
+  // Generate a version id for the new object.
+  new_object_metadata->regenerate_version_id();
+  new_object_metadata->set_oid(motr_writer->get_oid());
+  new_object_metadata->set_layout_id(layout_id);
+
   next();
   s3_log(S3_LOG_DEBUG, "", "Exiting\n");
 }
 
+void S3CopyObjectAction::create_object_failed() {
+  s3_log(S3_LOG_INFO, request_id, "Entering\n");
+  if (check_shutdown_and_rollback()) {
+    s3_log(S3_LOG_DEBUG, "", "Exiting\n");
+    return;
+  }
+  if (motr_writer->get_state() == S3MotrWiterOpState::exists) {
+    // collision_detected();
+  } else {
+    s3_copy_action_state = S3CopyObjectActionState::newObjOidCreationFailed;
+
+    if (motr_writer->get_state() == S3MotrWiterOpState::failed_to_launch) {
+      s3_log(S3_LOG_ERROR, request_id, "Create object failed.\n");
+      set_s3_error("ServiceUnavailable");
+    } else {
+      s3_log(S3_LOG_WARN, request_id, "Create object failed.\n");
+      // Any other error report failure.
+      set_s3_error("InternalError");
+    }
+    send_response_to_s3_client();
+  }
+  s3_log(S3_LOG_DEBUG, "", "Exiting\n");
+}
+
 // Copy source object to destination object
 void S3CopyObjectAction::copy_object() {
   s3_log(S3_LOG_INFO, request_id, "Entering\n");
-  read_object();              // read source object
-  initiate_data_streaming();  // write to destination object
-  next();
+
+  if (!content_length) {
+    s3_log(S3_LOG_DEBUG, request_id, "Source object is empty");
+    next();
+    return;
```

Comment:
do not add return statement, otherwise we will not get "Exiting" statement printed in logs.

Instead add the below section in else part of this if block.

Reply:
It seems to me that "Exiting" logs are non-informative in such cases.
Furthermore, long if/else blocks reduce readability.
Do you insist?